lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/service/VcStoringService.java

```diff
@@ -0,0 +1,216 @@
+            IpvSessionItem ipvSessionItem)
+            throws SqsException, CiPostMitigationsException, CiPutException, ParseException,
+                    JsonProcessingException {
+        criId = callbackRequest.getCredentialIssuerId();
+        ipAddress = callbackRequest.getIpAddress();
+        criOAuthSessionId = callbackRequest.getState();
+        ipvSessionId = callbackRequest.getIpvSessionId();
+        userId = clientOAuthSessionItem.getUserId();
+        govukSigninJourneyId = clientOAuthSessionItem.getGovukSigninJourneyId();
+        criConfig = credentialIssuerConfig;
+        componentId = configService.getSsmParameter(ConfigurationVariable.COMPONENT_ID);
+        this.ipvSessionItem = ipvSessionItem;
+
+        auditEventUser =
+                new AuditEventUser(
+                        userId, callbackRequest.getIpvSessionId(), govukSigninJourneyId, ipAddress);
+
+        var visitedCriDetails =
+                switch (vcResponse.getCredentialStatus()) {
+                    case CREATED -> storeCreatedVcs(vcResponse);
+                    case PENDING -> storePendingVcs(vcResponse);
+                };
+
+        ipvSessionItem.addVisitedCredentialIssuerDetails(visitedCriDetails);
+    }
+
+    private VisitedCredentialIssuerDetailsDto storePendingVcs(
+            VerifiableCredentialResponse vcResponse) throws JsonProcessingException, SqsException {
+        if (!vcResponse.getUserId().equals(userId)) {
+            throw new VerifiableCredentialResponseException(
+                    HTTPResponse.SC_SERVER_ERROR,
+                    FAILED_TO_VALIDATE_VERIFIABLE_CREDENTIAL_RESPONSE);
+        }
+
+        VerifiableCredentialResponseDto vcResponseDto =
+                VerifiableCredentialResponseDto.builder()
+                        .userId(userId)
+                        .credentialStatus(VerifiableCredentialStatus.PENDING.getStatus())
+                        .build();
+
+        criResponseService.persistCriResponse(
+                userId,
+                criId,
+                OBJECT_MAPPER.writeValueAsString(vcResponseDto),
+                criOAuthSessionId,
+                CriResponseService.STATUS_PENDING);
+
+        sendAuditEventForProcessedVcResponse(
+                CriResourceRetrievedType.PENDING.getType(), criId, auditEventUser);
+
+        return new VisitedCredentialIssuerDetailsDto(criId, componentId, false, null);
+    }
+
+    private VisitedCredentialIssuerDetailsDto storeCreatedVcs(
+            VerifiableCredentialResponse vcResponse)
+            throws SqsException, ParseException, JsonProcessingException, CiPutException,
+                    CiPostMitigationsException, VerifiableCredentialResponseException {
+        var vcs = vcResponse.getVerifiableCredentials();
+
+        var issuers = new HashSet<String>();
+        for (SignedJWT vc : vcs) {
+            verifiableCredentialJwtValidator.validate(vc, criConfig, userId);
```

Comment:
Do you think it would be reasonable to pull this check out of here, so we have a separate method (perhaps in the checking service) that validates the VCs before storage?

I don't think it's terrible to do it here, if inconvenient for some reason.